src/analyses/apron/relationAnalysis.apron.ml

```diff
@@ -635,13 +844,120 @@ struct
+                      if M.tracing then M.trace "malloc" "isPointerBeforeEnd: %a\n" ID.pretty isPointerBeforeEnd;
+                      ID.join  isPointerBeforeEnd st
+                    | _ -> ID.top ()
+                  )  a  (ID.bot ()) ;
+
+              | _ -> ID.top ()
+            in
+            if M.tracing then M.trace "malloc" "Query result: %a\n" VDQ.ID.pretty isBeforeEnd ;
+            (isAfterZero, isBeforeEnd)
+          | _ -> (VDQ.ID.top (),VDQ.ID.top ())
+        end
+      in
+      let relationEval = 
+        let i = match i with 
+          | `Lifted i -> i
+          | `Bot ->  IntDomain.IntDomTuple.top_of (Cilfacade.ptrdiff_ikind ())
+          | `Top -> IntDomain.IntDomTuple.top_of (Cilfacade.ptrdiff_ikind ()) 
+        in
+        begin match IntDomain.IntDomTuple.maximal i with 
+          | None -> (VDQ.ID.top (),VDQ.ID.top ())
+          | Some i ->  
+            let i = Z.to_int i in
+            begin match e with 
+              | Lval (Var v, _) -> (VDQ.ID.top (),VDQ.ID.top ())
+              | BinOp (binop, e1, e2, t) when binop = PlusPI || binop = MinusPI || binop = IndexPI -> 
+                let replaceBinop bi = match bi with
+                  | PlusPI | IndexPI -> PlusA
+                  | MinusPI -> MinusA
+                  | _ -> bi
+                in
+                let ptr_typ = typeOf e1 in
+                begin match ptr_typ with 
+                  | TPtr (t, _) -> 
+                    let sizeOfType = (bitsSizeOf t) / 8 in
+                    let e2Mult = BinOp (Mult, e2, integer sizeOfType, t) in
+                    let relExp = BinOp (replaceBinop binop, integer i, e2Mult, TInt (IInt ,[])) in
+
+                    let afterZero = Cilfacade.makeBinOp Le  Cil.zero relExp in
+                    if M.tracing then M.trace "OOB" "afterZero2: %a %a\n" d_exp e d_exp afterZero;
+                    let isAfterZero = eval_int afterZero (no_overflow ask afterZero) in
+                    if M.tracing then M.trace "OOB" "result: %a\n" ID.pretty isAfterZero;
+
+                    let isBeforeEnd = match ctx.ask (Queries.MayPointTo e) with 
+                      | a when not (Queries.AD.is_top a) -> 
+                        Queries.AD.fold ( fun (addr : AD.elt)  (st )   ->
+                            match addr with
+                            | Addr (v,o) -> 
+                              let allocatedLength = AllocSize.to_varinfo v in
+                              let pointerBeforeEnd = Cilfacade.makeBinOp Lt relExp (Lval (Var allocatedLength, NoOffset)) in
+                              if M.tracing then M.trace "OOB" "pointerBeforeEnd2: %a\n" d_exp pointerBeforeEnd;
+                              let isPointerBeforeEnd = eval_int pointerBeforeEnd (no_overflow ask pointerBeforeEnd) in
+                              if M.tracing then M.trace "OOB" "isPointerBeforeEnd2: %a\n" ID.pretty isPointerBeforeEnd;
+                              ID.join  isPointerBeforeEnd st
```

Comment:
After finding another bug concerning isAfterZero and extracting the anonymous function, the code can no longer be considered duplicate. If you think otherwise, please reopen the conversation.